### PR TITLE
[Torch] Weighted nodes search by metatypes in quantization algo

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

* Search of nodes with weights in quantization algorithm is implemented using metatypes instead of `target_model.get_weighted_original_graph_nodes()`

* New `PTModule` metatypes are introduced to represent operations wrapped by NNCF

* New attribute in `DynamicGraphNode` - `is_called_inside_nncf_module` is introduced. It used then during `NNCFGraph` building to distinguish operation which could be covered as NNCFLayer or registered by `register_module` but actually not covered as an NNCFLayer.

*  Subtypes now have valid attributes

* `replace_modules` is never wrapping nested NNCFModules: during initial module replacement it's working fine, but `replace_modules` on already wrapped model (inside `NNCFNetwork.temporary_clean_view` for example) was wrapping nested NNCF modules. 

### Reason for changes

* This changes allow to rely on metatypies during weighted node collection, as it's working in all other backends.

* PTOperatorMetatype isn't ambiguous now: for exmaple, `nn.Conv2d` module now has `PTModuleConv2dMetatype` metatype and `F.conv2d` has `PTConv2dMetatype` metatype. Both operation have `PTConv2dMetatype` on develop branch

*  `PTDepthwiseConvXdSubtype` previously have empty `module_to_function_names`, so `get_all_aliases` for subtypes was invalid

### Tickets

98133

### Tests

TODO:

- [x] Test new metatypes collection
- [x] Test user registered modules have valid metatypes
- [x] Test user registered modules are being searched as modules with wegihts

### CODE REVIEW

**Review by alexsu52 — COMMENTED** [review]
Could run e2e test for this PR?

File: nncf/torch/dynamic_graph/context.py [hunk 1]
```diff
@@ -129,7 +130,8 @@ def maybe_add_node(self,
                        tensor_metas: List[Optional[TensorMeta]],
                        op_address: OperationAddress,
                        module_attrs: BaseLayerAttributes = None,
-                       ignored_algorithms: List[str] = None) -> Optional[DynamicGraphNode]:
+                       ignored_algorithms: List[str] = None,
+                       is_called_inside_nncf_module: bool = False) -> Optional[DynamicGraphNode]:
```
Comment (ljaljushkin) [inline nncf/torch/dynamic_graph/context.py:134]: Is it supposed to be merged on top of #1428? I see this change in both PR, it's a bit confusing.

File: tests/torch/test_nncf_network.py [hunk 2]
```diff
@@ -493,22 +537,34 @@ def test_insertion_point_data_in_ip_nodes(self):
...
             PTMaxPool2dMetatype, PTTransposeMetatype, \
             PTConvTranspose2dMetatype, PTDepthwiseConv2dSubtype, PTAddMetatype, PTAvgPool2dMetatype, PTLinearMetatype
         ref_scope_vs_metatype_dict = {
             "/" + MODEL_INPUT_OP_NAME + "_0": PTInputNoopMetatype,
-            "ModelForMetatypeTesting/NNCFConv2d[conv_regular]/conv2d_0": PTConv2dMetatype,
-            "ModelForMetatypeTesting/NNCFBatchNorm2d[bn]/batch_norm_0": PTBatchNormMetatype,
+            "ModelForMetatypeTesting/NNCFConv2d[conv_regular]/conv2d_0":
+                PTConv2dMetatype.module_metatype,
+            "ModelForMetatypeTesting/NNCFBatchNorm2d[bn]/batch_norm_0":
+                PTBatchNormMetatype.module_metatype,
```
Comment (ljaljushkin) [inline tests/torch/test_nncf_network.py:548]: Shouldn't it be tested with registered user modules?
https://github.com/openvinotoolkit/nncf/pull/1430/files#r1048213174

**Review by ljaljushkin — CHANGES_REQUESTED** [review]
LGTM, but I see some tests are missing, and potentially it could lead to new changes. So can't approve now.

File: tests/torch/test_nncf_network.py (hunk 2 above)
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py:548]: `test_operator_metatype_marking` is testing static NNCF modules metatypes (from `NNCF_MODULES_DICT`) and 
`test_custom_module_registering` is all about custom module, it checks that modules are being wrapped and all metatypes are correct

File: nncf/torch/utils.py [hunk 3]
```diff
@@ -398,6 +398,8 @@ def get_model_device(model: torch.nn.Module) -> torch.device:
         device = torch.device('cpu')
     return device
 
+def warning_deprecated(msg):
+    warnings.warn(msg, DeprecationWarning)
```
Comment (daniil-lyakhov) [inline nncf/torch/utils.py:402]: TODO: remove

File: tests/torch/test_layer_attributes.py [hunk 4]
```diff
@@ -154,7 +153,7 @@ def __str__(self):
             groups=2,
             transpose=False,
             padding_values=(0, 0)),
-        metatype_cls=PTDepthwiseConv2dSubtype
```
Comment (ljaljushkin) [inline tests/torch/test_layer_attributes.py:157]: maybe it's worth checking subtype as well

Comment (daniil-lyakhov) [reply tests/torch/test_layer_attributes.py:157]: As it described [here](https://github.com/openvinotoolkit/nncf/pull/1428/files#diff-eaaa92a602e4cf9ff5c8ed3e761a1d66f4f2cf4f045e3c9e3c7b07902c769aaaR80) models that are just one module aren't replaced by nncf layers. So if module isn't replaced by `PTNNCFConv2d`, it has OP metatype, so it couldn't have `PTDepthwiseConv2d` metatype because it is now the suptype of `PTModuleConv2dMetatype`  and `PTModuleConv2dMetatype` is the subtype of `PTNNCFConv2d`.

Comment (daniil-lyakhov) [reply tests/torch/test_layer_attributes.py:157]: I believe that logic of metatypes is covered in `tests/torch/test_nncf_network.py`

File: nncf/torch/quantization/default_quantization.py [hunk 5]
```diff
@@ -60,6 +68,11 @@
     ],
     QuantizationTrait.OUTPUT_QUANTIZATION_AS_WEIGHTS: [
         operator_metatypes.PTEmbeddingMetatype,
-        operator_metatypes.PTEmbeddingBagMetatype
+        operator_metatypes.PTEmbeddingMetatype.module_metatype,
```
Comment (ljaljushkin) [inline nncf/torch/quantization/default_quantization.py:71]: You decided not to add module metatype automatically, didn't you?

Comment (daniil-lyakhov) [reply nncf/torch/quantization/default_quantization.py:71]: I overthink and decide that explicit typing is better

File: nncf/torch/quantization/default_quantization.py [hunk 6]
```diff
@@ -13,21 +13,27 @@
...
 # If there are no some metatypes it means that they are considered as QuantizationTrait.QuantizationAgnostic
 
 DEFAULT_PT_QUANT_TRAIT_TO_OP_DICT = {
     QuantizationTrait.INPUTS_QUANTIZABLE: [
         operator_metatypes.PTConv2dMetatype,
+        operator_metatypes.PTModuleConv2dMetatype,
         operator_metatypes.PTConv3dMetatype,
+        operator_metatypes.PTModuleConv3dMetatype,
         operator_metatypes.PTConvTranspose2dMetatype,
+        operator_metatypes.PTModuleConvTranspose2dMetatype,
```
Comment (ljaljushkin) [inline nncf/torch/quantization/default_quantization.py:30]: Any reason why you decided to introduce module metatypes? make it more explicit and aligned with ordinary metatypes?

**Review by ljaljushkin — COMMENTED** [review]
Did you manage to pass e2e tests?

File: nncf/torch/quantization/default_quantization.py (hunk 6 above)
Comment (daniil-lyakhov) [reply nncf/torch/quantization/default_quantization.py:30]: The main issue is that to find out which metatypes have module subtypes and which haven't you should know how the mapping between NNCFModules and metatypes. Yes, in case someone decide to introduce new NNCFLayer one should implement module metatype as well, but explicit typing helps to identify is metatype has module subtype or not

**Comment by daniil-lyakhov** [comment]
E2E test passed except `mobilenet_v2_imagenet_int4_int8`. It was confirmed on develop branch (build 151 and 152) that this model sporadically doesn't pass the test because expected accuracy drop is close to 1% border (0.97%) and due to GPU computation deviations sometimes accuracy drop is lager that 1%. So, actual E2E test for this pr is green, limits update for this model will be introduced in separate PR

**Review by ljaljushkin — APPROVED** [review]

**Comment by daniil-lyakhov** [comment]
run pytorch pre-commit tests

**Comment by daniil-lyakhov** [comment]
run pytorch pre-commit tests

File: tests/torch/test_pytorch_patch.py [hunk 7]
```diff
@@ -1,24 +1,23 @@
...
 from nncf.torch.dynamic_graph.patch_pytorch import MagicFunctionsToPatch
 from nncf.torch.graph.operator_metatypes import PT_OPERATOR_METATYPES
 from nncf.torch.dynamic_graph.context import TracingContext
 from nncf.torch.dynamic_graph.trace_tensor import TracedTensor
 from nncf.torch.dynamic_graph.trace_tensor import TensorMeta
 
 
-def test_metatype_op_names_do_not_intersect():
-    function_name_to_operator_metatype = defaultdict(list)
+def test_get_all_aliaces_is_valid():
```
Comment (vshampor) [inline tests/torch/test_pytorch_patch.py:10]: `aliases`

**Review by vshampor — CHANGES_REQUESTED** [review]
A small grammar change and we are good to go here.

File: tests/torch/test_pytorch_patch.py (hunk 7 above)
Comment (daniil-lyakhov) [reply tests/torch/test_pytorch_patch.py]: Done!

**Review by kshpv — APPROVED** [review]

**Review by vshampor — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
LGTM
